ngraph/test/type_prop/unary_base.cpp

```diff
@@ -0,0 +1,178 @@
+};
+
+template <class Op>
+void dynamic_rank_input_shape(InputType type)
+{
+    const auto param = std::make_shared<op::Parameter>(type, PartialShape::dynamic());
+    const auto op = std::make_shared<Op>(param);
+    ASSERT_TRUE(op->get_output_partial_shape(0).same_scheme(PartialShape::dynamic()));
+};
+
+class BasicParamInference : public testing::WithParamInterface<UnaryParams>, public testing::Test
+{
+public:
+    InputType input_type;
+    Shape input_shape;
+
+    static std::string getTestCaseName(testing::TestParamInfo<UnaryParams> obj)
+    {
+        InputType type;
+        Shape shape;
+        std::ostringstream result;
+        std::tie(type, shape) = obj.param;
+        result << type << "_";
+        result << shape;
+        auto test_name = result.str();
+        std::replace_if(
+            test_name.begin(), test_name.end(), [](char c) { return !std::isalnum(c); }, '_');
+        test_name.erase(test_name.end() - 1);
+        return test_name;
+    }
+
+    void SetUp()
+    {
+        const auto& type = std::get<0>(GetParam());
+        const auto& shape = std::get<1>(GetParam());
+
+        input_type = type;
+        input_shape = shape;
+    }
+};
+
+class IncompatibleInputType : public testing::WithParamInterface<UnaryParams>, public testing::Test
+{
+public:
+    InputType input_type;
+    Shape input_shape;
+
+    static std::string getTestCaseName(testing::TestParamInfo<UnaryParams> obj)
+    {
+        InputType type;
+        Shape shape;
+        std::ostringstream result;
+        std::tie(type, shape) = obj.param;
+        result << type << "_";
+        result << shape;
+        auto test_name = result.str();
+        std::replace_if(
+            test_name.begin(), test_name.end(), [](char c) { return !std::isalnum(c); }, '_');
+        test_name.erase(test_name.end() - 1);
+        return test_name;
+    }
+
+    void SetUp()
+    {
+        const auto& type = std::get<0>(GetParam());
+        const auto& shape = std::get<1>(GetParam());
+
+        input_type = type;
+        input_shape = shape;
+    }
+};
+
+class DynamicRankInputShape : public testing::WithParamInterface<std::tuple<InputType>>,
+                              public testing::Test
+{
+public:
+    InputType input_type;
+
+    static std::string getTestCaseName(testing::TestParamInfo<std::tuple<InputType>> obj)
+    {
+        InputType type;
+        std::ostringstream result;
+        type = std::get<0>(obj.param);
+        result << type;
+        return result.str();
+    }
+
+    void SetUp()
+    {
+        const auto& type = std::get<0>(GetParam());
+        input_type = type;
+    }
+};
+
+TEST_P(BasicParamInference, acos)
+{
+    basic_param_inference<op::Acos>(input_type, input_shape);
+}
+
+TEST_P(IncompatibleInputType, acos)
+{
+    incompatible_input_type<op::Acos>(input_type, input_shape);
+}
+
+TEST_P(DynamicRankInputShape, acos)
+{
+    dynamic_rank_input_shape<op::Acos>(input_type);
+}
```

Comment:
Did you try to parametrize further your tests with operation type (e.g. op::Acos) ? There is [Typed Tests](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#typed-tests) feature available, but I didn't try it myself yet.

Reply:
Did not try, but if is preferred I can check that